sklearn/neighbors/base.py

```diff
@@ -331,7 +340,11 @@ class from an array representing our data set and ask who's
 
         if X is not None:
             query_is_train = False
-            X = check_array(X, accept_sparse='csr')
+            if self.effective_metric_ in _MASKED_SUPPORTED_METRICS:
```

Comment:
I think we should not do any checking here if method is 'brute' (though we then may need to use `_num_samples(X)` instead of `X.shape` below).

Reply:
Not sure I follow. Could you explain a bit please?

Reply:
If method is 'brute' then any requirements that the metric has on the data can be checked by the metric (or by `pairwise_distances`). However, in the BinaryTree cases, we need to do checking here. Yet, if we don't check it, we might have a list of lists or something, not necessarily an object with `.shape`. So we need to use `_num_samples` which is designed to handle the case of getting the first axis length of unvalidated array-likes.

Reply:
okay so that would mean the check can be avoided for brute both in the if-true and the else-true case correct? i.e. regardless of whether self.effective_metric_ in _MASKED_SUPPORTED_METRICS?

Reply:
I've not double checked but I think so.

Reply:
hmm this is causing some issues:

line 1375, in pairwise_distances
    if(np.any(_get_mask(X, missing_values).sum(axis=1) == X.shape[1])):
AttributeError: 'list' object has no attribute 'shape'